File: payment_generator.R
set.seed(42)

# Name and role data used to build worker records dynamically
first_names_male <- c("James","John","Robert","Michael","William","David",
  "Richard","Joseph","Thomas","Charles","Christopher","Daniel",
  "Matthew","Anthony","Mark","Donald","Steven","Paul","Andrew","Kenneth")

first_names_female <- c("Mary","Patricia","Jennifer","Linda","Barbara",
  "Elizabeth","Susan","Jessica","Sarah","Karen","Lisa","Nancy",
  "Betty","Margaret","Sandra","Ashley","Dorothy","Kimberly","Emily","Donna")

last_names <- c("Smith","Johnson","Williams","Brown","Jones","Garcia",
  "Miller","Davis","Rodriguez","Martinez","Hernandez","Lopez",
  "Gonzalez","Wilson","Anderson","Thomas","Taylor","Moore","Jackson","Martin")

roles <- c("Site Engineer","Project Manager","Foreman","Electrician",
  "Plumber","Carpenter","Mason","Welder","Safety Officer","Equipment Operator")

# Dynamically build a data frame of 420 worker records
count    <- 420
genders  <- sample(c("Male","Female"), count, replace = TRUE)
salaries <- round(runif(count, 5000, 35000), 2)
firsts   <- ifelse(genders == "Male",
                   sample(first_names_male,   count, replace = TRUE),
                   sample(first_names_female, count, replace = TRUE))

workers <- data.frame(
  id     = sprintf("HCC-%04d", 1:count),
  name   = paste(firsts, sample(last_names, count, replace = TRUE)),
  gender = genders,
  role   = sample(roles, count, replace = TRUE),
  salary = salaries,
  stringsAsFactors = FALSE
)

cat(sprintf("Generated %d workers.\n\n", nrow(workers)))

# Loop through all workers and generate a payment slip for each one
payment_slips <- list()

for (i in seq_len(nrow(workers))) {
  tryCatch({
    w      <- workers[i, ]
    salary <- w$salary
    gender <- w$gender

    # Validate salary is numeric
    if (!is.numeric(salary)) stop(paste("Invalid salary for", w$id))

    # Salary must not be negative
    if (salary < 0) stop(paste("Negative salary for", w$id))

    # Assign employee level based on salary and gender
    # A5-F is checked first — a qualifying female takes this level over A1
    if (salary > 7500 && salary < 30000 && gender == "Female") {
      level <- "A5-F"
    } else if (salary > 10000 && salary < 20000) {
      level <- "A1"
    } else {
      level <- "N/A"
    }

    # Build the payment slip for this worker
    payment_slips[[i]] <- data.frame(
      id = w$id, name = w$name, gender = gender,
      role = w$role, salary = salary, level = level,
      stringsAsFactors = FALSE
    )

  }, error = function(e) {
    # Log the error and continue processing remaining workers
    cat(sprintf("Error on row %d: %s\n", i, conditionMessage(e)))
  })
}

# Combine all slips into a single data frame
slips <- do.call(rbind, Filter(Negate(is.null), payment_slips))

# Print a preview of the first 10 payment slips
cat("Sample payment slips:\n")
cat(strrep("-", 70), "\n")
for (i in 1:min(10, nrow(slips))) {
  s <- slips[i, ]
  cat(sprintf("[%s] %-28s | %-6s | $%10.2f | Level: %s\n",
              s$id, s$name, s$gender, s$salary, s$level))
}

# Print overall payroll summary
level_counts <- table(slips$level)
cat(sprintf("\nTotal slips: %d\n", nrow(slips)))
cat(sprintf("Total payroll: $%s\n", format(sum(slips$salary), big.mark=",")))
cat("\nEmployee Level Distribution:\n")
for (lvl in sort(names(level_counts))) {
  cat(sprintf("  %s: %d workers\n", lvl, level_counts[[lvl]]))
}

File: payment_generator.py
import random

random.seed(42)

first_names_male = ["James", "John", "Robert", "Michael", "William", "David",
    "Richard", "Joseph", "Thomas", "Charles", "Christopher", "Daniel",
    "Matthew", "Anthony", "Mark", "Donald", "Steven", "Paul", "Andrew", "Kenneth"]

first_names_female = ["Mary", "Patricia", "Jennifer", "Linda", "Barbara",
    "Elizabeth", "Susan", "Jessica", "Sarah", "Karen", "Lisa", "Nancy",
    "Betty", "Margaret", "Sandra", "Ashley", "Dorothy", "Kimberly", "Emily", "Donna"]

last_names = ["Smith", "Johnson", "Williams", "Brown", "Jones", "Garcia",
    "Miller", "Davis", "Rodriguez", "Martinez", "Hernandez", "Lopez",
    "Gonzalez", "Wilson", "Anderson", "Thomas", "Taylor", "Moore", "Jackson", "Martin"]

roles = ["Site Engineer", "Project Manager", "Foreman", "Electrician",
    "Plumber", "Carpenter", "Mason", "Welder", "Safety Officer", "Equipment Operator"]

# Step 2: Dynamically create a list of 420 workers
workers = []
for i in range(1, 421):
    gender = random.choice(["Male", "Female"])
    first = random.choice(first_names_male if gender == "Male" else first_names_female)
    worker = {
        "id": f"HCC-{i:04d}",
        "name": f"{first} {random.choice(last_names)}",
        "gender": gender,
        "role": random.choice(roles),
        "salary": round(random.uniform(5000, 35000), 2),
    }
    workers.append(worker)

print(f"Generated {len(workers)} workers.\n")

# Step 3 & 4: For loop with conditional statements + Step 5: Exception handling
payment_slips = []

for worker in workers:
    try:
        salary = worker["salary"]
        gender = worker["gender"]

        if not isinstance(salary, (int, float)):
            raise TypeError(f"Invalid salary type for {worker['id']}")
        if salary < 0:
            raise ValueError(f"Negative salary for {worker['id']}")

        # Conditional 1 & 2 — A5-F checked first (takes priority)
        if salary > 7500 and salary < 30000 and gender == "Female":
            level = "A5-F"
        elif salary > 10000 and salary < 20000:
            level = "A1"
        else:
            level = "N/A"

        slip = {
            "id":     worker["id"],
            "name":   worker["name"],
            "gender": gender,
            "role":   worker["role"],
            "salary": salary,
            "level":  level,
        }
        payment_slips.append(slip)

    except (KeyError, TypeError, ValueError) as e:
        print(f"Error processing worker: {e}")

# Print first 10 slips
print("Sample payment slips:")
print("-" * 70)
for slip in payment_slips[:10]:
    print(f"[{slip['id']}] {slip['name']:<28} | {slip['gender']:<6} | "
          f"${slip['salary']:>10,.2f} | Level: {slip['level']}")

# Summary
from collections import Counter
levels = Counter(s["level"] for s in payment_slips)
print(f"\nTotal slips generated: {len(payment_slips)}")
print(f"Total payroll: ${sum(s['salary'] for s in payment_slips):,.2f}")
print("\nEmployee Level Distribution:")
for lvl, count in sorted(levels.items()):
    print(f"  {lvl}: {count} workers")